src/rinterface_extra.c

```diff
@@ -2576,33 +2576,59 @@ SEXP R_igraph_0orvector_complex_to_SEXP(const igraph_vector_complex_t *v) {
 }
 
 SEXP R_igraph_matrix_to_SEXP(const igraph_matrix_t *m) {
-
   SEXP result, dim;
+  const igraph_integer_t nrow = igraph_matrix_nrow(m);
+  const igraph_integer_t ncol = igraph_matrix_ncol(m);
+
+  /* Assuming that this function is called in a context where
+   * igraph_error() does not return. */
+  if (nrow > INT_MAX || ncol > INT_MAX) {
+    igraph_errorf("igraph returned a matrix of size %" IGRAPH_PRId " by %" IGRAPH_PRId ". "
+                  "R does not support matrices with more than %d rows or columns.",
+                  __FILE__, __LINE__, IGRAPH_FAILURE,
+                  nrow, ncol, INT_MAX);
+  }
```

Comment:
This could be extracted into a function, say, `R_igraph_check_matrix_dims(igraph_integer_t nrow, igraph_integer_t ncol)` . This function becomes testable, e.g., through an R wrapper that we export just for testing. (Overkill, perhaps.)

We could also go one step further and make this function sensitive to a global switch (one that's perhaps only active in debug mode). The global switch would change the definition of `INT_MAX` for this function so that we can even test that (some) functions that return a matrix actually go through this code path. (Overkill, perhaps.)

But these two steps combined allow creating a version of igraph that doesn't require allocating hundreds of gigabytes to check the functionality.

No action needed here, but for the future, should we strive to take testability into account when implementing?

Reply:
> should we strive to take testability into account when implementing

This PR fixes a bug that manifests itself in difficult to test scenarios. Nevertheless it is a bug that should be fixed. I did verify the fix manually.

> This function becomes testable, e.g., through an R wrapper that we export just for testing.

The function would take arguments of type `igraph_integer_t`. This type is not supported by R. Using `double` makes things complicated and increases the chance of mistakes (not ever `igraph_integer_t` is representable as a `double`).

So I'm not sure how to do this in a productive way.

Reply:
It's in the merge queue, thanks for working on this.

Happy to elaborate if this situation occurs on an issue from https://github.com/igraph/rigraph/milestone/14.